src/shared/file-record.js

```diff
@@ -0,0 +1,516 @@
+/// <reference path="./file-record-types.js" />
+
+import _fs from 'fs';
+import _path from 'path';
+
+/**
+ * @enum {FileRecord.Action}
+ */
+const Action = {
+  /** @type {'parallel'} */
+  PARALLEL: 'parallel',
+  /** @type {'sequence'} */
+  SEQUENCE: 'sequence',
+  /** @type {'makeDirectory'} */
+  MAKE_DIRECTORY: 'makeDirectory',
+  /** @type {'writeFile'} */
+  WRITE_FILE: 'writeFile',
+  /** @type {'unlinkEntry'} */
+  UNLINK_ENTRY: 'unlinkEntry',
+};
+
+const host = createHost();
+
+export { host, createHost, isDirectory, isFile, Action };
+
+function isDirectory(record) {
+  return Boolean(record && 'entries' in record && record.entries);
+}
+
+function isFile(record) {
+  return Boolean(record && 'buffer' in record && record.buffer);
+}
+
+function createHost({ fs = _fs.promises, path = _path } = {}) {
```

Comment:
I think I understand what you are getting at but I'm not sure how this isn't OOP.

I think I opted for this style because the records are data structures with no methods. As well the built host doesn't have internal state between the set of exposed functions. That `fs` and `path` are similar basic object collections of functions probably weighed into it as well.